File: answers/other/answers/lintcode/Search a 2D Matrix.cpp
class Solution {
public:
    /**
     * @param matrix, a list of lists of integers
     * @param target, an integer
     * @return a boolean, indicate whether matrix contains target
     */
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        // write your code here
        if (0 == matrix.size() || 0 == matrix[0].size()) {
        	return false;
        }
        size_t left = 0;
        size_t right = matrix.size() - 1;
        while (left <= right && right < matrix.size()) {
        	size_t middle = (left + right) / 2;
        	if (target < matrix[middle][0]) {
        		right = middle - 1;
        	}
        	else if (target >= matrix[middle][0] && (middle + 1 >= matrix.size() || target < matrix[middle + 1][0])) {
        		break;
        	}
        	else {
        		left = middle + 1;
        	}
        }
        size_t row = (left + right) / 2;
        left = 0;
        right = matrix[row].size() - 1;
        while (left <= right && right < matrix[row].size()) {
        	size_t middle = (left + right) / 2;
        	if (target == matrix[row][middle]) {
        		return true;
        	}
        	if (target < matrix[row][middle]) {
        		right = middle - 1;
        	}
        	else {
        		left = middle + 1;
        	}
        }
        return false;
    }
};

File: answers/leetcode/Search a 2D Matrix/Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }
        int row = 0;
        int left = 0;
        int right = matrix.length - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (matrix[middle][0] == target) {
                return true;
            } else if (matrix[middle][0] > target) {
                right--;
            } else if (matrix[middle][0] < target && middle + 1 < matrix.length && matrix[middle + 1][0] <= target) {
                left++;
            } else {
                row = middle;
                break;
            }
        }
        left = 0;
        right = matrix[row].length - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            int value = matrix[row][middle];
            if (value > target) {
                right--;
            } else if (value < target) {
                left++;
            } else {
                return true;
            }
        }
        return false;
    }
}